source/opt/ccp_pass.cpp

```diff
@@ -0,0 +1,267 @@
+      *dest_bb = nullptr;
+      return SSAPropagator::kVarying;
+    }
+
+    // Get the constant value for the selector from the value table. Use it to
+    // decide which branch will be taken.
+    uint32_t select_val_id = it->second;
+    const analysis::Constant* c =
+        const_mgr_->FindDeclaredConstant(select_val_id);
+    assert(c && "Expected to find a constant declaration for a known value.");
+    const analysis::IntConstant* val = c->AsIntConstant();
+
+    // Start assuming that the selector will take the default value;
+    dest_label = instr->GetSingleWordOperand(1);
+    for (uint32_t i = 2; i < instr->NumOperands(); i += 2) {
+      if (val->words()[0] == instr->GetSingleWordOperand(i)) {
+        dest_label = instr->GetSingleWordOperand(i + 1);
+        break;
+      }
+    }
+  }
+
+  assert(dest_label && "Destination label should be set at this point.");
+  *dest_bb = context()->cfg()->block(dest_label);
+  return SSAPropagator::kInteresting;
+}
+
+SSAPropagator::PropStatus CCPPass::VisitInstruction(ir::Instruction* instr,
+                                                    ir::BasicBlock** dest_bb) {
+  *dest_bb = nullptr;
+  if (instr->opcode() == SpvOpPhi) {
+    return VisitPhi(instr);
+  } else if (instr->IsBranch()) {
+    return VisitBranch(instr, dest_bb);
+  } else if (instr->result_id()) {
+    return VisitAssignment(instr);
+  }
+  return SSAPropagator::kVarying;
+}
+
+bool CCPPass::ReplaceValues() {
+  bool retval = false;
+  for (const auto& it : values_) {
+    uint32_t id = it.first;
+    uint32_t cst_id = it.second;
+    if (id != cst_id) {
+      context()->get_def_use_mgr()->ForEachUser(
+          id, [this, cst_id, id, &retval](ir::Instruction* instr) {
+            instr->ForEachInOperand([id, cst_id, &retval](uint32_t* id_p) {
+              if (id == *id_p) {
+                *id_p = cst_id;
+                retval = true;
+              }
+            });
+          });
+    }
+  }
+  return retval;
+}
+
+bool CCPPass::PropagateConstants(ir::Function* fp) {
+  const auto visit_fn = [this](ir::Instruction* instr,
+                               ir::BasicBlock** dest_bb) {
+    return VisitInstruction(instr, dest_bb);
+  };
+
+  InsertPhiInstructions(fp);
+  propagator_ =
+      std::unique_ptr<SSAPropagator>(new SSAPropagator(context(), visit_fn));
+  if (propagator_->Run(fp)) {
+    return ReplaceValues();
+  }
+
+  return false;
+}
+
+void CCPPass::Initialize(ir::IRContext* c) {
+  InitializeProcessing(c);
+
+  const_mgr_ = context()->get_constant_mgr();
+
+  // Populate the constant table with values from constant declarations in the
+  // module.  The values of each OpConstant declaration is the identity
+  // assignment (i.e., each constant is its own value).
+  for (const auto& inst : c->module()->GetConstants()) {
+    values_[inst->result_id()] = inst->result_id();
+    const_mgr_->MapConstantToInst(const_mgr_->GetConstantFromInst(inst), inst);
```

Comment:
Add ConstantManager::MapInst and merge these functions together?

Reply:
MapInst?

Reply:
I suggest merging "const_mgr_->MapConstantToInst(const_mgr_->GetConstantFromInst(inst), inst)" together into a single new function maybe called MapInst. Since the input since to rely on other constant manager information it might be convenient to provide the single function too.